zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,309 @@
+//! The compress circuit implementation.
+
+use crate::evm_circuit::util::from_bytes;
+use crate::util::{Expr, SubCircuitConfig};
+use bit_vec::BitVec;
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector, TableColumn},
+    poly::Rotation,
+};
+use lazy_static::lazy_static;
+use std::marker::PhantomData;
+
+lazy_static! {
+    // TODO: maybe fixed huffman table will get from witness
+    static ref FIXED_HUFFMAN_TABLE: [BitVec; 256] = [(); 256].map(|_| BitVec::new());
+}
+
+mod combine {
+    const LEFT_SHIFT_8BITS: u64 = 256;
+
+    use crate::util::Expr;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+    pub(crate) fn expr<F: Field>(parts: [Expression<F>; 32]) -> Expression<F> {
+        parts
+            .into_iter()
+            .fold(0.expr(), |acc, part| acc * F::from(LEFT_SHIFT_8BITS) + part)
+    }
+}
+
+mod compress {
+    use super::FIXED_HUFFMAN_TABLE;
+    use crate::evm_circuit::util::from_bytes;
+    use eth_types::Field;
+
+    pub(crate) fn value<F: Field>(src: F) -> [F; 32] {
+        src.to_repr().map(|item| {
+            let code = FIXED_HUFFMAN_TABLE[item as usize].to_bytes();
+            from_bytes::value::<F>(&code)
+        })
+    }
+}
+
+mod rlc {
+    use crate::util::Expr;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(
+        parts: [Expression<F>; 32],
+        randomness: Expression<F>,
+    ) -> Expression<F> {
+        parts
+            .into_iter()
+            .rev()
+            .fold(0.expr(), |acc, part| acc * randomness.clone() + part)
+    }
+
+    pub(crate) fn value<F: Field>(src: [F; 32], randomness: F) -> F {
+        src.into_iter()
+            .rev()
+            .fold(F::zero(), |acc, part| acc * randomness + part)
+    }
+}
```

Comment:
Any reason this is redefined here instead of reusing the one evm_circuit/util.rs?

Reply:
Sorry, no need to redefine.